plugins/Login/Security/BruteForceDetection.php

```diff
@@ -0,0 +1,134 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+namespace Piwik\Plugins\Login\Security;
+
+use Piwik\Common;
+use Piwik\Date;
+use Piwik\Db;
+use Piwik\Plugins\Login\SystemSettings;
+
+class BruteForceDetection {
+
+    private $minutesTimeRange;
+    private $maxLogAttempts;
+
+    private $table = 'brute_force_log';
+    private $tablePrefixed = '';
+
+    /**
+     * @var SystemSettings
+     */
+    private $settings;
+
+    public function __construct(SystemSettings $systemSettings)
+    {
+        $this->tablePrefixed = Common::prefixTable($this->table);
+        $this->settings = $systemSettings;
+        $this->minutesTimeRange = $systemSettings->loginAttemptsTimeRange->getValue();
+        $this->maxLogAttempts = $systemSettings->maxFailedLoginsPerMinutes->getValue();
+    }
+
+    public function isEnabled()
+    {
+        return $this->settings->enableBruteForceDetection->getValue();
+    }
+
+    public function addFailedLoginAttempt($ipAddress)
+    {
+        $now = $this->getNow()->getDatetime();
+        $db = Db::get();
+        $db->query('INSERT INTO '.$this->tablePrefixed.' SET ip_address = ?, attempted_at = ?', array($ipAddress, $now));
```

Comment:
mysqlism here. don't think it's a big deal since matomo and mysql are pretty intertwined at this point, just letting you know in case you missed it.

Reply:
will update 👍